apple/internal/resource_actions/metals.bzl

```diff
@@ -0,0 +1,107 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Metal related actions."""
+
+load(
+    "@build_bazel_apple_support//lib:apple_support.bzl",
+    "apple_support",
+)
+load(
+    "@bazel_skylib//lib:paths.bzl",
+    "paths",
+)
+
+def _metal_apple_target_triple(ctx):
+    """Returns a Metal target triple string for an Apple platform.
+
+    Args:
+        ctx: The target's context.
+    Returns:
+        A target triple string describing the platform.
+    """
+    platform = ctx.fragments.apple.single_arch_platform
+    xcode_config = ctx.attr._xcode_config[apple_common.XcodeVersionConfig]
+    target_os_version = xcode_config.minimum_os_for_platform_type(
+        platform.platform_type,
+    )
+
+    platform_string = str(platform.platform_type)
+    if platform_string == "macos":
+        platform_string = "macosx"
+
+    environment = "" if platform.is_device else "-simulator"
+
+    return "air64-apple-{platform}{version}{environment}".format(
+        environment = environment,
+        platform = platform_string,
+        version = target_os_version,
+    )
+
+def compile_metals(ctx, input_files, output_file):
+    """Creates actions that compile .metal files into a single .metallib file.
+
+    Args:
+        ctx: The target's rule context.
+        input_files: The input metal files.
+        output_file: The output metallib file.
+    """
+    air_files = []
+    target = _metal_apple_target_triple(ctx)
+
+    if not input_files:
+        fail("Input .metal files can't be empty")
+
+    # Compile each .metal file into a single .air file
+    for input_metal in input_files:
+        air_file = ctx.actions.declare_file(
+            paths.replace_extension(input_metal.basename, ".air"),
+        )
+        air_files.append(air_file)
+
+        args = ctx.actions.args()
+        args.add("metal")
+        args.add("-c")
+        args.add("-target", target)
+        args.add("-ffast-math")
+        args.add("-o", air_file.path)
+        args.add(input_metal.path)
+
+        apple_support.run(
+            ctx,
+            executable = "/usr/bin/xcrun",
+            inputs = [input_metal],
+            outputs = [air_file],
+            arguments = [args],
+            mnemonic = "MetalCompile",
+        )
+
+    # Compile .air files into a single .metallib file, which stores the Metal
+    # library
+    args = ctx.actions.args()
+    args.add("metallib")
+    args.add("-o", output_file.path)
+    args.add_all([
```

Comment:
`args.add_all(air_files)` should work here, without having to expand `air_files` at all